middlets.ts

```diff
@@ -0,0 +1,165 @@
+import express, { Request, Response, NextFunction } from 'express';
+import axios from 'axios';
+import asyncRetry from 'async-retry';
+import jwt from 'jsonwebtoken';
+import mongoose, { ConnectOptions } from 'mongoose';
+
+const app = express();
+const port = 3000;
+const SECRET_KEY = 'gauss626';
+const MONGO_URL = 'mongodb://127.0.0.1:27017/mydatabase';
+const COLLECTION_NAME = 'log-collection';
+
+// Definir el esquema de los logs
+const logSchema = new mongoose.Schema({
+  resourceAccessed: String,
+  userId: String,
+  dateCreated: {
+    type: Date,
+    default: Date.now
+  }
+});
+
+// Modelo de los logs
+const Log = mongoose.model('Log', logSchema);
+
+// // Middleware para registrar los logs en la colección de MongoDB
+// app.use(async (req: Request, res: Response, next: NextFunction) => {
+//   const { userId } = req.headers;
+//   const logEntry = {
+//     resourceAccessed: req.path,
+//     userId,
+//   };
+
+//   try {
+//     await Log.create(logEntry);
+//   } catch (error) {
+//     console.error('Error al guardar el log en MongoDB:', error);
+//   }
+
+//   next();
+// });
+// Middleware para registrar los logs en la colección de MongoDB
+app.use(async (req: Request, res: Response, next: NextFunction) => {
+  const { userId } = req.headers;
+  const logEntry = {
+    resourceAccessed: req.path,
+    userId: userId || '', // Capturar el valor de userId del encabezado de la solicitud
+  };
+
+  try {
+    const log = await Log.create(logEntry);
+    const formattedLog = log.toObject(); // Transformar el objeto a un formato plano
+    console.log('Registro de log guardado:', formattedLog);
+  } catch (error) {
+    console.error('Error al guardar el log en MongoDB:', error);
+  }
+
+  next();
+});
+
+
+
+// Middleware para verificar la autorización del token JWT
+const checkAuthorization = (req: Request, res: Response, next: NextFunction) => {
+  const { userId, userRole } = req.headers;
+
+  if (!userId || userRole !== 'ADMIN') {
+    return res.sendStatus(401); // Unauthorized
+  }
+
+  next();
+};
+
+// Transformar los datos del usuario según el formato requerido
+const transformUser = (user: any) => ({
+  userId: user.id,
+  name: user.name,
+  email: user.email,
+  postId: "",
+  title: "",
+  body: ""
+});
+
+// Ruta GET /users
+app.get('/users', async (req: Request, res: Response) => {
+  try {
+    const response = await asyncRetry(() => axios.get('https://jsonplaceholder.typicode.com/users'));
```

Comment:
Let's extract this logic to a handler which should call its controller class.

Move the axios calls of the api to a file called `jsonPlaceholderService` and create methods for each of the API calls that you are doing.